src/MPS.jl

```diff
@@ -0,0 +1,335 @@
+
+Returns an [`MPS`](@ref) of `n` sites whose tensors are initialized to COPY-tensors.
+
+# Keyword Arguments
+
+  - `physdim` The physical or output dimension of each site. Defaults to 2.
+  - `maxdim` The maximum bond dimension. Defaults to `physdim^(n ÷ 2)`.
+"""
+function MPS(::typeof(identity), n::Integer; physdim=2, maxdim=physdim^(n ÷ 2))
+    # Create bond dimensions until the middle of the MPS considering maxdim
+    virtualdims = min.(maxdim, physdim .^ (1:(n ÷ 2)))
+
+    # Complete the bond dimensions of the other half of the MPS
+    virtualdims = vcat(virtualdims, virtualdims[(isodd(n) ? end : end - 1):-1:1])
+
+    # Create each site dimensions in default order (:o, :l, :r)
+    arraysdims = [[physdim, virtualdims[1]]]
+    append!(arraysdims, [[physdim, virtualdims[i], virtualdims[i + 1]] for i in 1:(length(virtualdims) - 1)])
+    push!(arraysdims, [physdim, virtualdims[end]])
+
+    # Create the MPS with copy-tensors according to the tensors dimensions
+    return MPS(
+        map(arraysdims) do arrdims
+            arr = zeros(ComplexF64, arrdims...)
+            deltas = [fill(i, length(arrdims)) for i in 1:physdim]
+            broadcast(delta -> arr[delta...] = 1.0, deltas)
+            arr
+        end,
+    )
+end
+
+function Base.convert(::Type{MPS}, tn::Product)
+    @assert socket(tn) == State()
+
+    arrs::Vector{Array} = arrays(tn)
+    arrs[1] = reshape(arrs[1], size(arrs[1])..., 1)
+    arrs[end] = reshape(arrs[end], size(arrs[end])..., 1)
+    map!(@view(arrs[2:(end - 1)]), @view(arrs[2:(end - 1)])) do arr
+        reshape(arr, size(arr)..., 1, 1)
+    end
+
+    return MPS(arrs)
+end
+
+Base.adjoint(tn::MPS) = MPS(adjoint(Ansatz(tn)), form(tn))
+
+# TODO different input/output physical dims
+# TODO let choose the orthogonality center
+function Base.rand(rng::Random.AbstractRNG, ::Type{MPS}; n, maxdim, eltype=Float64, physdim=2)
+    p = physdim
+    T = eltype
+    χ = maxdim
+
+    arrays::Vector{AbstractArray{T,N} where {N}} = map(1:n) do i
+        χl, χr = let after_mid = i > n ÷ 2, i = (n + 1 - abs(2i - n - 1)) ÷ 2
+            χl = min(χ, p^(i - 1))
+            χr = min(χ, p^i)
+
+            # swap bond dims after mid and handle midpoint for odd-length MPS
+            (isodd(n) && i == n ÷ 2 + 1) ? (χl, χl) : (after_mid ? (χr, χl) : (χl, χr))
+        end
+
+        # orthogonalize by QR factorization
+        F = lq!(rand(rng, T, χl, p * χr))
+
+        reshape(Matrix(F.Q), χl, p, χr)
+    end
+
+    # reshape boundary sites
+    arrays[1] = reshape(arrays[1], p, p)
+    arrays[n] = reshape(arrays[n], p, p)
+
+    return MPS(arrays; order=(:l, :o, :r))
+end
+
+# TODO deprecate contract(; between) and generalize it to AbstractAnsatz
+"""
+    Tenet.contract!(tn::MPS; between=(site1, site2), direction::Symbol = :left, delete_Λ = true)
+
+For a given [`MPS`](@ref) tensor network, contracts the singular values Λ between two sites `site1` and `site2`.
```

Comment:
contracts the tensors at site `site1` and `site2` with the SV between them ? que pasa si site1 y site2 are not adjacent?

Reply:
one option is that if `site1` and `site2` are not neighbors (i.e. there is no edge connecting them in the lattice), then this should error
but for some stuff i'm interested in contracting non-contiguous tensors, so might only print a warning